kernel/src/engine/arrow_data.rs

```diff
@@ -143,184 +109,160 @@ impl EngineMap for MapArray {
+        // Make sure the caller passed the correct number of column names
+        let leaf_types = visitor.selected_column_names_and_types().1;
+        if leaf_types.len() != leaf_columns.len() {
+            return Err(Error::MissingColumn(format!(
+                "Visitor expected {} column names, but caller passed {}",
+                leaf_types.len(), leaf_columns.len()
+            )).with_backtrace());
+        }
+
+        // Collect the names of all leaf columns we want to extract, along with their parents, to
+        // guide our depth-first extraction. If the list contains any non-leaf, duplicate, or
+        // missing column references, the extracted column list will be too short (error out below).
+        let mut mask = HashSet::new();
+        for column in leaf_columns {
+            for i in 0..column.len() {
+                mask.insert(&column[..i+1]);
+            }
+        }
+        debug!("Column mask for selected columns {leaf_columns:?} is {mask:#?}");
+
+        let mut getters = vec![];
+        Self::extract_columns(&mut vec![], &mut getters, leaf_types, &mask, &self.data)?;
+        if getters.len() != leaf_columns.len() {
+            return Err(Error::MissingColumn(
+                format!(
+                    "Visitor expected {} leaf columns, but only {} were found in the data",
+                    leaf_columns.len(), getters.len()
+                )
+            ));
+        }
+        visitor.visit(self.len(), &getters)
+    }
+}
+
+impl ArrowEngineData {
+    fn extract_columns<'a>(
+        path: &mut Vec<String>,
+        getters: &mut Vec<&'a dyn GetData<'a>>,
+        leaf_types: &[DataType],
+        column_mask: &HashSet<&[String]>,
+        data: &'a dyn ProvidesColumnsAndFields,
     ) -> DeltaResult<()> {
-        for field in schema.fields() {
-            let col = array
-                .and_then(|a| a.column_by_name(&field.name))
-                .filter(|a| *a.data_type() != ArrowDataType::Null);
-            // Note: if col is None we have either:
-            //   a) encountered a column that is all nulls or,
-            //   b) recursed into a optional struct that was null. In this case, array.is_none() is
-            //      true and we don't need to check field nullability, because we assume all fields
-            //      of a nullable struct can be null
-            // So below if the field is allowed to be null, OR array.is_none() we push that,
-            // otherwise we error out.
-            if let Some(col) = col {
-                Self::extract_column(out_col_array, field, col)?;
-            } else if array.is_none() || field.is_nullable() {
-                if let DataType::Struct(inner_struct) = field.data_type() {
-                    Self::extract_columns_from_array(out_col_array, inner_struct.as_ref(), None)?;
+        for (column, field) in data.columns().iter().zip(data.fields()) {
```

Comment:
this assumes that the data will always contain a column for everything requested. the previous code tried to handle the case where the column was not present in the underlying data but was nullable, so it just put in a null getter.

I _think_ this is okay because we always read with the full schema so arrow will create a `null` column in the output data, but I haven't thought through everything yet. Just wanted to check if you considered this?

Reply:
> Just wanted to check if you considered this?

From the PR description:
> NOTE: As part of this change, the visitor framework no longer tolerates missing columns in the EngineData being visited. The parquet reader should have already null-filled those.

If the visitor requests a column that doesn't exist in the logical schema, I don't think it's correct to just null-fill it. We had that problem with the parquet row group skipping not so long ago...

Reply:
Right, sorry I missed that. Makes sense that we rely on the parquet reader to fill cols missing in the data. I'll note that this _also_ gets used over data that comes from the json reader. So we need to be very clear in our descriptions of what's expected from those.  I tested a bit and our json reads does seem to null fill everything, but I believe that is a detail of the arrow reader that may change. They [do say](https://docs.rs/arrow-json/53.3.0/arrow_json/reader/struct.ReaderBuilder.html) they will ignore columns not in the schema, but not that they will null fill those that are present in the schema but missing in the data.

Reply:
Good point about documenting the requirement for json readers to null-pad missing data... but it's already baked in. For example, most commits don't include metadata or protocol changes, and most _tables_ have _never_ seen a set_transaction action. And yet we happily visit those columns during log replay, expecting them to be properly null-filled. Even the simple P&M query would fail if json readers didn't allow that.